sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/enhancedfanout/EFOShardSubscribersPool.java

```diff
@@ -196,33 +200,45 @@ public void start(Iterable<ShardCheckpoint> checkpoints) {
    */
   @Nullable
   KinesisRecord getNextRecord() throws IOException {
-    if (current == null && eventQueue.isEmpty()) {
-      if (subscriptionError == null) {
-        return null;
-      } else {
-        throw new IOException(subscriptionError);
+    while (true) {
+      if (!isStopped && subscriptionError != null) {
+        // Stop the pool to cancel all subscribers and prevent new subscriptions.
+        // Doing this as part of getNextRecord() avoids concurrent access to the state map and
+        // prevents any related issues.
+        stop();
```

Comment:
Let's move cancellation of subscribers from the error handler into `getNextRecord` to prevent race conditions.

Reply:
Note: Despite stopping the pool, the reader can continue to consume all pending records in the queue until the exception is finally thrown.